pallets/inflation/src/lib.rs

```diff
@@ -0,0 +1,638 @@
+
+            ActiveInflationConfig::<T>::put(config.clone());
+
+            Self::deposit_event(Event::<T>::InflationConfigurationForceChanged { config });
+
+            Ok(().into())
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        /// Used to check if inflation recalculation is supposed to happen on the next block.
+        ///
+        /// This will be true even if recalculation is overdue, e.g. it should have happened in the current or older block.
+        fn is_recalculation_in_next_block(
+            now: BlockNumber,
+            config: &InflationConfiguration,
+        ) -> bool {
+            config.recalculation_block.saturating_sub(now) <= 1
+        }
+
+        /// Payout block rewards to the beneficiaries.
+        ///
+        /// Return the total amount issued.
+        fn payout_block_rewards() -> Balance {
+            let config = ActiveInflationConfig::<T>::get();
+
+            let collator_amount = T::Currency::issue(config.collator_reward_per_block);
+            let treasury_amount = T::Currency::issue(config.treasury_reward_per_block);
+
+            T::PayoutPerBlock::collators(collator_amount);
+            T::PayoutPerBlock::treasury(treasury_amount);
+
+            config.collator_reward_per_block + config.treasury_reward_per_block
+        }
+
+        /// Recalculates the inflation based on the total issuance & inflation parameters.
+        ///
+        /// Returns the new inflation configuration.
+        pub(crate) fn recalculate_inflation(now: BlockNumber) -> InflationConfiguration {
+            let params = InflationParams::<T>::get();
+            let total_issuance = T::Currency::total_issuance();
+
+            // 1. Calculate maximum emission over the period before the next recalculation.
+            let max_emission = params.max_inflation_rate * total_issuance;
+            let issuance_safety_cap = total_issuance.saturating_add(max_emission);
+
+            // 2. Calculate distribution of max emission between different purposes.
+            let treasury_emission = params.treasury_part * max_emission;
+            let collators_emission = params.collators_part * max_emission;
+            let dapps_emission = params.dapps_part * max_emission;
+            let base_stakers_emission = params.base_stakers_part * max_emission;
+            let adjustable_stakers_emission = params.adjustable_stakers_part * max_emission;
+            let bonus_emission = params.bonus_part * max_emission;
+
+            // 3. Calculate concrete rewards per block, era or period
+
+            // 3.0 Convert all 'per cycle' values to the correct type (Balance).
+            // Also include a safety check that none of the values is zero since this would cause a division by zero.
+            // The configuration & integration tests must ensure this never happens, so the following code is just an additional safety measure.
+            let blocks_per_cycle = match T::CycleConfiguration::blocks_per_cycle() {
```

Comment:
I think this ensure the safety check of values not being zero:
```
let blocks_per_cycle = Balance::from(T::CycleConfiguration::blocks_per_cycle().max(1));
let build_and_earn_eras_per_cycle =
    Balance::from(T::CycleConfiguration::build_and_earn_eras_per_cycle().max(1));
let periods_per_cycle =
    Balance::from(T::CycleConfiguration::periods_per_cycle().max(1));
```

Reply:
Looks much better, thanks!